pkg/controller/dynamicupgradeableconditionscontroller.go

```diff
@@ -0,0 +1,300 @@
+package controller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+	"sync"
+
+	semver "github.com/blang/semver/v4"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/cluster-olm-operator/pkg/clients"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	helmclient "github.com/operator-framework/helm-operator-plugins/pkg/client"
+	storage "github.com/operator-framework/helm-operator-plugins/pkg/storage"
+	ocv1alpha1 "github.com/operator-framework/operator-controller/api/v1alpha1"
+	"github.com/operator-framework/operator-registry/alpha/property"
+	helm "helm.sh/helm/v3/pkg/storage"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	incompatibleOperatorsInstalled = "IncompatibleOperatorsInstalled"
+	MaxOpenShiftVersionProperty    = "olm.maxOpenShiftVersion"
+	OwnerKindKey                   = "olm.operatorframework.io/owner-kind"
+	OwnerNameKey                   = "olm.operatorframework.io/owner-name"
+	PackageNameKey                 = "olm.operatorframework.io/package-name"
+	BundleNameKey                  = "olm.operatorframework.io/bundle-name"
+	BundleVersionKey               = "olm.operatorframework.io/bundle-version"
+)
+
+type dynamicUpgradeableConditionController struct {
+	kubeclient     kubernetes.Interface
+	informer       informers.GenericInformer
+	client         dynamic.Interface
+	name           string
+	operatorClient *clients.OperatorClient
+	prefixes       []string
+}
+
+type ChartMetadata struct {
+	Metadata struct {
+		Annotations map[string]string `yaml:"annotations"`
+	} `yaml:"metadata"`
```

Comment:
Is there a way this could reasonably be pulled out into a separate type? This notation threw me off at first glance.

Maybe something like:
```go
type Metadata struct {
    Annotations map[string]string `yaml:"annotations"`
}

type Chart struct {
    Metadata Metadata `yaml:"metadata"`
}
```

Reply:
This notation is somewhat common when declaring one-off structs for parsing like we are in this case. I actually prefer the current PR's setup in that regard.

Though I do notice that `ChartMetadata` is exported. We should unexport that.